### ISSUE TITLE
Session State Transition After FINISH_RSP

### ISSUE THREAD

**Issue by bokyungseo1218** [DMTF/libspdm#2150]
Do you know any reason why libspdm code checks capabilities as below during libspdm_build_response of FINISH_RSP.
https://github.com/DMTF/libspdm/blob/main/library/spdm_responder_lib/libspdm_rsp_receive_send.c#L772
```c++
if (session_id != NULL) {
        switch (request_response_code) {
        case SPDM_FINISH_RSP:
            if (!libspdm_is_capabilities_flag_supported(
                    context, false,
                    SPDM_GET_CAPABILITIES_REQUEST_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP,
                    SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP)) {
                libspdm_set_session_state(
                    context, *session_id,
                    LIBSPDM_SESSION_STATE_ESTABLISHED);
            }
            break;
        case SPDM_PSK_FINISH_RSP:
            libspdm_set_session_state(context, *session_id,
                                      LIBSPDM_SESSION_STATE_ESTABLISHED);
            break;
``` 
As I know, session state have to change from LIBSPDM_SESSION_STATE_HANDSHAKING to LIBSPDM_SESSION_STATE_ESTABLISHED

**Comment by jyao1** [comment]
This is session case.

Non-session case is at: https://github.com/DMTF/libspdm/blob/main/library/spdm_responder_lib/libspdm_rsp_receive_send.c#L821

**Comment by bokyungseo1218** [comment]
@jyao1  Thank you for your quick reply
After finish_rsp, libspdm_session_state has to be changed from  LIBSPDM_SESSION_STATE_HANDSHAKING to LIBSPDM_SESSION_STATE_ESTABLISHED?
(regardless of whether SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP is supported or not)

https://github.com/DMTF/libspdm/blob/6674aa87c5a3b6991e25e22fcd1bf44e2a111331/include/library/spdm_secured_message_lib.h#L29
```
typedef enum {
    /* Before send KEY_EXCHANGE/PSK_EXCHANGE or after END_SESSION */
    LIBSPDM_SESSION_STATE_NOT_STARTED,

    /* After send KEY_EXCHANGE, before send FINISH */
    LIBSPDM_SESSION_STATE_HANDSHAKING,

    /* After send FINISH, before END_SESSION */
    LIBSPDM_SESSION_STATE_ESTABLISHED,

    /* MAX */
    LIBSPDM_SESSION_STATE_MAX
} libspdm_session_state_t;
```

**Comment by jyao1** [comment]
> After finish_rsp, libspdm_session_state has to be changed from LIBSPDM_SESSION_STATE_HANDSHAKING to LIBSPDM_SESSION_STATE_ESTABLISHED?
> (regardless of whether SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP is supported or not)

right.

**Comment by bokyungseo1218** [comment]
Thank you. 

If so, shouldn't the session state have to be changed regardless of capabilities at below codes
https://github.com/DMTF/libspdm/blob/main/library/spdm_responder_lib/libspdm_rsp_receive_send.c#L768
```c
    if (session_id != NULL) {
        switch (request_response_code) {
        case SPDM_FINISH_RSP:
            if (!libspdm_is_capabilities_flag_supported(
                    context, false,
                    SPDM_GET_CAPABILITIES_REQUEST_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP,
                    SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP)) {
                libspdm_set_session_state(
                    context, *session_id,
                    LIBSPDM_SESSION_STATE_ESTABLISHED);
            }
            break;
``` 
and 
https://github.com/DMTF/libspdm/blob/main/library/spdm_responder_lib/libspdm_rsp_receive_send.c#L817
```c
        switch (request_response_code) {
        case SPDM_FINISH_RSP:
            if (libspdm_is_capabilities_flag_supported(
                    context, false,
                    SPDM_GET_CAPABILITIES_REQUEST_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP,
                    SPDM_GET_CAPABILITIES_RESPONSE_FLAGS_HANDSHAKE_IN_THE_CLEAR_CAP)) {
                libspdm_set_session_state(
                    context,
                    context->latest_session_id,
                    LIBSPDM_SESSION_STATE_ESTABLISHED);
            }
            break;
``` 

I would like to know why libspdm checks capabilities to change session state at SPDM_FINISH_RSP

**Comment by xiangfeima** [comment]
This comes from the specification, see SPDM 1.2.1 paragraph 478:
> The Requester and Responder pair can support two modes of handshakes. If HANDSHAKE_IN_THE_CLEAR_CAP is set
in both the Requester and the Responder all SPDM messages exchanged during the Session Handshake Phase are
sent in the clear (outside of a secure session). Otherwise both the Requester and the Responder use encryption and/
or message authentication during the Session Handshake Phase using the Handshake secret derived at the
completion of KEY_EXCHANGE_RSP message for subsequent message communication until FINISH_RSP message
completion.

**Comment by bokyungseo1218** [comment]
@xiangfeima 
thank you for reply

if HANDSHAKE_IN_THE CLEARE_CAP is set, 
when/how session state is changed from LIBSPDM_SESSION_STATE_HANDSHAKING to LIBSPDM_SESSION_STATE_ESTABLISHED after FINISH_RSP message?

**Comment by xiangfeima** [comment]
> @xiangfeima thank you for reply
> 
> if HANDSHAKE_IN_THE CLEARE_CAP is set, when/how session state is changed from LIBSPDM_SESSION_STATE_HANDSHAKING to LIBSPDM_SESSION_STATE_ESTABLISHED after FINISH_RSP message?

That goes to the else branch
https://github.com/DMTF/libspdm/blob/eb9ab5c71dadce9c94d6a864179b4da460ec9199/library/spdm_responder_lib/libspdm_rsp_receive_send.c#L814-L826

And in this case (HANDSHAKE_IN_THE CLEARE_CAP is set) the FINISH_RSP message is `in the clear (outside of session)`, the session_id for determing state of which session is to be changed is from previously cached `context->latest_session_id`.

**Comment by bokyungseo1218** [comment]
@xiangfeima 
Thank you so much

If so, 
In case of HANDSHAKE_IN_THE CLEARE_CAP is set, shouldn't the value below table be changed from "Prohibited" to "Partial Allowed"?
Table 6 — SPDM request and response messages validity

| Request | Response | Session | Outside of a session| Session phases|
|--------|--------|--------|--------|--------|
| FINISH | FINISH_RSP | Allowed | Prohibited | Session Handshake|

**Comment by xiangfeima** [comment]
It is confirmed to be bug in specification, I believe that has been raised to SPDM work group, see https://github.com/DMTF/libspdm/pull/1778#issuecomment-1463052334
@steven-bellock is there update on this?

**Comment by bokyungseo1218** [comment]
@xiangfeima 
Thank you

**Comment by steven-bellock** [comment]
The fix did not make it into 1.3, but should be in 1.2.2, whenever that gets published.

**Comment by xiangfeima** [comment]
Thanks @steven-bellock for the info.
